doc/good-practices.mld

```diff
@@ -0,0 +1,405 @@
+definition. When called, the driver is called with the [-check] and
+[-check-locations] flags, [ppxlib] makes it is a requirement that locations follow
+some rules in order to accept the rewriting, as it will check that some
+invariants are respected.
+
+{2 The Invariants}
+
+The invariants are as follows:
+
+- AST nodes are requested to be well-nested WRT locations
+- the locations of "sibling" AST nodes should not overlap
+
+This is required for Merlin to behave properly.
+
+Indeed, for almost any query directed at Merlin, it will need to inspect the
+context around the user's cursor to give an answer that makes sense. And the
+only input it has to do that is the cursor’s position in the buffer.
+The handling of most queries starts by traversing the AST, using the
+locations of nodes to select the right branch. (1) is necessary to avoid
+discarding subtrees too early, (2) is used to avoid Merlin making arbitrary
+choices (if you ask for the type under the cursor, and there seems to be two
+things under the cursor, Merlin will need to pick one).
+
+{2 Guidelines for Writing Well-Behaved PPXs}
+
+It's obviously not always (indeed rarely) possible to mint new locations
+when manipulating the AST.
+
+The intended way to deal with locations is this:
+
+- AST nodes that exist in the source should keep their original location
+- new nodes should be given a "ghost" location (i.e.,
+  [{ some_loc with loc_ghost = true }]) to indicate that the node doesn't
+  exist in the sources.
+
+Both the new check and Merlin will happily traverse these ghost nodes as if
+they didn't exist. Note: this comes into play when deciding which nodes are
+"siblings," for instance, if your AST is:
+
+{v
+  A (B1(C, D),
+     B2(X, Y))
+v}
+
+but [B2] has a ghost location, then [B1], [X] and [Y] are considered
+siblings.
+
+Additionally, there is an attribute [\[@merlin.hide\]] that you can add on
+nodes to tell Merlin (and the check) to ignore this node and all of its
+children. Some helpers for this are provided in {{!Ppxlib.Merlin_helpers}[Merlin_helpers]}.
+
+{1 Testing Your PPX}
+
+This section is not yet written. You can refer to {{:https://tarides.com/blog/2019-05-09-an-introduction-to-ocaml-ppx-ecosystem#testing-your-ppx}this blog post}, or contribute to the [ppxlib] documentation by opening a pull request in the {{:https://github.com/ocaml-ppx/ppxlib/}repository}.
+
+{1 Integrating With Merlin}
+
+This section is not yet written. You can contribute to the [ppxlib] documentation by opening a pull request in the {{:https://github.com/ocaml-ppx/ppxlib/}repository}.
+
+{1 Migrate From OMP}
+
+This section is not yet written. You can contribute to the [ppxlib] documentation by opening a pull request in the {{:https://github.com/ocaml-ppx/ppxlib/}repository}.
+
+{1 Migrate From [Ppx_traverse]}
+
+This section is not yet written. You can contribute to the [ppxlib] documentation by opening a pull request in the {{:https://github.com/ocaml-ppx/ppxlib/}repository}.
```

Comment:
Oh, interesting! You're mentioning Ppx_traverse and not mentioning Camlp4/Camlp5. Where have you come across Ppx_traverse?

Reply:
The short history in this context is:

`ppx_traverse` is one of the preprocessing libraries that was used at Jane Street before `ppxlib`. Jane Street has fully shifted to `ppxlib` in everything preprocessing related.

Parallel to Jane Street's development, the OCaml community has shifted from Camlp4/Camlp5 (i.e. extending the OCaml parser itself) over writing compiler version specific PPXs just with `compiler-libs`, over writing cross-compiler compatible PPXs with `compiler-libs`+`OMP1`, to writing the PPXs we know today with `ppxlib`.

To be honest, I'd remove `Ppx_traverse`.

We _could_ mention Camlp5/Camlp4 and/or `cppo`. According to the [OCaml 2022 user survey](https://ocaml-sf.org/docs/2022/ocaml-user-survey-2022.pdf), 6,6% of people use a cpp-style conditional-compilation directives for preprocessing in OCaml and around 5% still use Camlp4/5.

Reply:
`ppx_traverse` is not an alternative to `ppxlib`, it's a `deriving` clause usable via `ppxlib`. It's used to generate the code for `ast_traverse.ml`, for example, and we use it for other things at Jane Street.

Reply:
Thanks for the context, @ceastlund! It's good to know that it's still used. The reason why I thought `ppxlib` comes from `ppx_traverse` (among many others) is the following sentence in our [history md](https://github.com/ocaml-ppx/ppxlib/blob/main/HISTORY.md):

> This repository was originally a merge of several ppx-related repositories, namely:
>  -   [ppx_ast](https://github.com/janestreet/ppx_ast);
> -  [ppx_core](https://github.com/janestreet/ppx_core);
>  -  [ppx_driver](https://github.com/janestreet/ppx_driver);
>  -  [ppx_metaquot](https://github.com/janestreet/ppx_metaquot);
>  -  [ppx_traverse](https://github.com/janestreet/ppx_traverse);
>  -  [ppx_traverse_builtins](https://github.com/janestreet/ppx_traverse_builtins);
>  -  [ppx_type_conv](https://github.com/janestreet/ppx_type_conv).

If you could give us more context about this, I think that would be very helpful for all of us!